src/dict.h

```diff
@@ -55,55 +55,54 @@ typedef struct dictEntry {
         int64_t s64;
         double d;
     } v;
-    struct dictEntry *next;
 } dictEntry;
 
 typedef struct dictType {
-    uint64_t (*hashFunction)(const void *key);
+    uint64_t (*hashFunction)(const void *key); /* TODO: add level? */
     void *(*keyDup)(void *privdata, const void *key);
     void *(*valDup)(void *privdata, const void *obj);
     int (*keyCompare)(void *privdata, const void *key1, const void *key2);
     void (*keyDestructor)(void *privdata, void *key);
     void (*valDestructor)(void *privdata, void *obj);
-    int (*expandAllowed)(size_t moreMem, double usedRatio);
+    int (*expandAllowed)(size_t moreMem, double usedRatio); /* DELETEME */
 } dictType;
 
-/* This is our hash table structure. Every dictionary has two of this as we
- * implement incremental rehashing, for the old to the new table. */
-typedef struct dictht {
-    dictEntry **table;
-    unsigned long size;
-    unsigned long sizemask;
-    unsigned long used;
-} dictht;
+/* The internal dict structure: a HAMT (see dict.c for details). */
+
+union dictNode; /* leaf or internal node */
+
+/* Internal node (same size as dictEntry) */
+typedef struct dictSubNode {
+    uint32_t bitmap;  /* exitence of potential child */
+    uint32_t leafmap; /* child is 1=leaf or 0=subnode */
+    union dictNode *children;
+} dictSubNode;
+
+/* The parent node contains a 'leafmap', which has a bit for each child to flag
+ * it as 1 = entry or 0 = subnode. */
+union dictNode {
+    dictEntry entry;
+    dictSubNode sub;
+};
```

Comment:
Yes, agree, I'll move as much as possible into the .c file, after more urgent issues have been solved. Since exposing dictEntry pointers is risky, as you say, it would be good to change the API to avoid exposing it and make dictEntry opaque too. Not sure if it's feasible or not.